### PR TITLE
In case of error from bloop shout down bsp server

### PR DESCRIPTION
**Pull request by romanowski**

### CODE REVIEW

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala [hunk 1]
```diff
@@ -20,11 +23,29 @@ class BspServer(
...
+
   private var extraDependencySources: Seq[os.Path] = Nil
   def setExtraDependencySources(sourceJars: Seq[os.Path]): Unit = {
     extraDependencySources = sourceJars
   }
 
+  // Can we accept some errors in some circumstances?
+  override protected def onFatalError(throwable: Throwable): Unit = {
+    val sw = new StringWriter()
+    throwable.printStackTrace(new PrintWriter(sw))
```
Comment (alexarchambault) [inline modules/build/src/main/scala/scala/build/bsp/BspServer.scala:38]: I think this discards the cause exceptions, which are often helpful. Better loop on the causes, like done [here](https://github.com/VirtusLab/scala-cli/blob/5d714dceaf4aa8049a4cc778725f8c47f7e3403f/modules/build/src/main/scala/scala/build/Build.scala#L809-L816).

**Review by alexarchambault — COMMENTED** [review]
When calling `.handleAsync(fatalExceptionHandler)`, I'm wondering if we could pass the method name (`buildShutdown`, `buildTargetCleanCache`, …) to `fatalExceptionHandler`,  and maybe also the method input params if it has any. These could be passed to `onFatalError`, that could print them to help debugging issues later on.

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala (hunk 1 above)
Comment (romanowski) [reply modules/build/src/main/scala/scala/build/bsp/BspServer.scala:38]: Are you sure? I run some simple test and it prints nicely nested exceptions (I am using `printStackTrace` like 10 years now :)

```
⋊> ~/w/P/scala-cli-pg bat errors.scala                                                                                                         10:50:36
   1   │ @main def errors = 
   2   │  try
   3   │    try ???
   4   │    catch 
   5   │      case e: Throwable => throw RuntimeException(e)
   6   │  catch
   7   │    case e => e.printStackTrace()
⋊> ~/w/P/scala-cli-pg scala-cli errors.scala                                                                                                   10:50:40
java.lang.RuntimeException: scala.NotImplementedError: an implementation is missing
	at errors$package$.errors(errors.scala:5)
	at errors.main(errors.scala:1)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at scala.cli.runner.Runner$.main(Runner.scala:14)
	at scala.cli.runner.Runner.main(Runner.scala)
Caused by: scala.NotImplementedError: an implementation is missing
	at scala.Predef$.$qmark$qmark$qmark(Predef.scala:344)
	at errors$package$.errors(errors.scala:3)
	... 7 more

```

**Comment by romanowski** [comment]
> I'm wondering if we could pass the method name (...)

That is a good idea, let me add it

**Comment by romanowski** [comment]
Logs looks like this:

```
2022.01.21 13:05:48 ERROR Fatal error has occured within bloop bsp server, method: buildTargetCompile, with params: CompileParams [
  targets = ArrayList (
    BuildTargetIdentifier [
      uri = "file:/home/krzysiek/workspace/scala-cli/.pg/b/.scala/?id=project_d476da5045"
    ]
  )
  originId = null
  arguments = null
]. Shutting down the server:
 org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.io.IOException: Broken pipe
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:72)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.request(RemoteEndpoint.java:161)
	at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:91)
	at jdk.proxy2/jdk.proxy2.$Proxy6.buildTargetCompile(Unknown Source)
	at scala.build.bsp.BuildServerForwardStubs.buildTargetCompile(BuildServerForwardStubs.scala:39)
	at scala.build.bsp.BuildServerForwardStubs.buildTargetCompile$(BuildServerForwardStubs.scala:38)
	at scala.build.bsp.BspServer.super$buildTargetCompile(BspServer.scala:157)
	at scala.build.bsp.BspServer.$anonfun$buildTargetCompile$1(BspServer.scala:157)
	at scala.build.bsp.BspImpl.$anonfun$compile$2(BspImpl.scala:237)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:132)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:97)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:53)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:532)
	at java.base/java.nio.channels.Channels.writeFullyImpl(Channels.java:74)
	at java.base/java.nio.channels.Channels.writeFully(Channels.java:93)
	at java.base/java.nio.channels.Channels$1.write(Channels.java:171)
	at java.base/java.io.OutputStream.write(OutputStream.java:127)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:67)
	... 14 more

2022.01.21 13:05:48 INFO  tracing is enabled: /home/krzysiek/.cache/metals/bsp.trace.json
2022.01.21 13:05:53 INFO  Connected to Build server: scala-cli v0.0.10-SNAPSHOT
```

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala (hunk 1 above)
Comment (alexarchambault) [reply modules/build/src/main/scala/scala/build/bsp/BspServer.scala:38]: Oh, I didn't know that, seems that method is more evolved than it looks!

**Comment by alexarchambault** [comment]
Looks like the BSP connection with Bloop was closed… Maybe Bloop exited somehow?

**Comment by romanowski** [comment]
Yeah, I have killed it to test if logging works :)

**Review by alexarchambault — APPROVED** [review]

**Pull request merged**